pol-core/bscript/compiler/astbuilder/CompoundStatementBuilder.cpp

```diff
@@ -87,6 +98,62 @@ std::unique_ptr<Block> CompoundStatementBuilder::block( EscriptParser::BlockCont
   return std::make_unique<Block>( location_for( *ctx ), std::move( statements ) );
 }
 
+std::unique_ptr<CaseStatement> CompoundStatementBuilder::case_statement(
+    EscriptParser::CaseStatementContext* ctx )
+{
+  // possibilities:
+  // case (expr)
+  //  4:
+  //  "xz":
+  //  CONSTANT:
+  //  default:
+  auto loc = location_for( *ctx );
+  std::string case_label;
+  if ( auto statement_label = ctx->statementLabel() )
+    case_label = text( statement_label->IDENTIFIER() );
+
+  auto determinant = expression( ctx->expression() );
+
+  std::vector<std::unique_ptr<CaseDispatchGroup>> dispatch_groups;
+
+  for ( auto group : ctx->switchBlockStatementGroup() )
+  {
+    std::vector<std::unique_ptr<Node>> selectors;
+
+    for ( auto group_label : group->switchLabel() )
+    {
+      if ( auto integer_literal = group_label->integerLiteral() )
+      {
+        selectors.push_back( integer_value( integer_literal ) );
+      }
+      else if ( auto identifier = group_label->IDENTIFIER() )
+      {
+        selectors.push_back(
+            std::make_unique<Identifier>( location_for( *identifier ), text( identifier ) ) );
+      }
+      else if ( auto string_literal = group_label->STRING_LITERAL() )
+      {
+        selectors.push_back( string_value( string_literal ) );
+      }
+      else if ( auto dflt = group_label->DEFAULT() )
+      {
+        selectors.push_back(
```

Comment:
Do we want to at least warn if more than a default is defined?

Reply:
Ah thanks, good catch.  Yes, but in the semantic analyzer, and for all of the selector types.

Reply:
Cool! I'm now curious how many duplicate ints/strings are out there.